src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs

```diff
@@ -1436,7 +1433,30 @@ internal DirectoryResponse ConstructResponse(int messageId, LdapOperation operat
                 needAbandon = false;
             }
 
-            int error = LdapPal.GetResultFromAsyncOperation(_ldapHandle, messageId, (int)resultType, timeout, ref ldapResult);
+            int error;
+            if (sync)
+            {
+                error = LdapPal.GetResultFromAsyncOperation(_ldapHandle, messageId, (int)resultType, timeout, ref ldapResult);
+            }
+            else
+            {
+                // Given that we will be polling every millisecond, we need a stopwatch to make sure we track the timeout
```

Comment:
Probably worth a comment about polling because there's no callback-based function to use.